File: src/894. All Possible Full Binary Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {number} n
 * @return {TreeNode[]}
 */

function allPossibleFBT(n) {
  const memo = {}
  return collectFBT(n)
  function collectFBT(n) {
    const results = []
    if (memo[n]) return memo[n]
    if (n % 2 === 0) return results
    if (n === 1) {
      results.push(new TreeNode(0))
      return (memo[n] = results)
    }
    for (let i = 1; i < n; i += 2) {
      const leftTrees = allPossibleFBT(i)
      const rightTrees = allPossibleFBT(n - 1 - i)
      for (let leftTree of leftTrees)
        for (let rightTree of rightTrees)
          results.push(new TreeNode(0, leftTree, rightTree))
    }
    return (memo[n] = results)
  }
}

File: src/894. All Possible Full Binary Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int n) {
        if (n % 2 == 0) {
            return {};
        }

        vector<vector<TreeNode*>> dp(n + 1);
        dp[1].push_back(new TreeNode(0));

        for (int i = 3; i <= n; i += 2) {
            for (int left = 1; left < i - 1; left += 2) {
                const int right = i - 1 - left;
                for (auto& leftChildren : dp[left]) {
                    for (auto& rightChildren : dp[right]) {
                        TreeNode* node = new TreeNode(0, leftChildren, rightChildren);
                        dp[i].push_back(node);
                    }
                }
            }
        }

        return dp[n];
    }
};